crates/sui-framework/packages/sui-framework/sources/random.move

```diff
@@ -122,4 +122,116 @@ module sui::random {
     ) {
         update_randomness_state(self, new_round, new_bytes, ctx);
     }
+
+
+    /// Unique randomness generator, derived from the global randomness.
+    struct RandomGenerator has drop {
+        seed: vector<u8>,
+        counter: u64,
+        buffer: vector<u8>,
+    }
+
+    const RAND_OUTPUT_LEN: u16 = 32;
+    const MAX_RANDOM_BYTES: u16 = RAND_OUTPUT_LEN * 100;
+
+    /// Create a generator.
+    public fun new_generator(r: &Random, ctx: &mut TxContext): RandomGenerator {
+        let inner = load_inner(r);
+        // TODO[crypto]: should blake be used here instead?
+        let seed = hmac_sha3_256(&inner.random_bytes,&to_bytes(fresh_object_address(ctx)));
+        RandomGenerator { seed, counter: 0, buffer: vector::empty() }
+    }
+
+    // TODO[move]: Should any of the following functions be implemented as native functions to save gas?
+
+    // Fill the buffer with 32 random bytes.
+    fun fill_buffer(g: &mut RandomGenerator) {
+        // TODO[crypto]: should blake be used here instead?
+        vector::append(&mut g.buffer, hmac_sha3_256(&g.seed, &bcs::to_bytes(&g.counter)));
+        g.counter = g.counter + 1;
+    }
+
+    /// Derive n random bytes.
+    public fun bytes(g: &mut RandomGenerator, num_of_bytes: u16): vector<u8> {
+        // TODO[move]: should we have this limit or just leave it to gas limit?
+        assert!(num_of_bytes <= MAX_RANDOM_BYTES, ETooManyBytes);
+        let result = vector::empty();
+        let num_of_derived_bytes = 0;
+        while (num_of_derived_bytes < num_of_bytes) {
+            if (vector::length(&g.buffer) == 0) {
+                fill_buffer(g);
+            };
+            vector::push_back(&mut result, vector::pop_back(&mut g.buffer));
+            num_of_derived_bytes = num_of_derived_bytes + 1;
+        };
+        result
+    }
+
+    // Helper function that extracts the given number of bytes from the random generator and returns it as u256.
+    // Assumes that the caller has already checked that num_of_bytes is valid.
+    fun u256_from_bytes(g: &mut RandomGenerator, num_of_bytes: u16): u256 {
```

Comment:
I guess `num_of_bytes` could be a `u8` right? I mean max value seems to be 32, no?
Having said so and considering it is a private function we could use a u64 and avoid the "cast" in the loop

Reply:
updated to use u64